<a href="https://colab.research.google.com/github/lalitpandey02/PythonNotebooks/blob/main/Career_Wednesday_18_Data_Manipulation_Assessment_Solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center><img src="https://github.com/insaid2018/Term-1/blob/master/Images/INSAID_Full%20Logo.png?raw=true" width="240" height="100" /></center>

**<center><h1>Career Wednesday - 18</h3></center>**

---
# **Data Manipulation Assessment**
---

  - ## [**Question 1**](#Section1)<br>
  - ## [**Question 2**](#Section2)<br>
  - ## [**Question 3**](#Section3)<br>

In [ ]:
#Installing Evaluation Script
!wget -q https://gitlab.com/academics3/insaid-data/-/raw/main/Career%20Wednesday%20Repo/17th%20August%202022:%20Data%20Manipulation%20Assessment/Evaluation.py

In [ ]:
from Evaluation import *

<a name = Section1></a>

---
# **Data Manipulation:**
---

In [ ]:
# Packages
import pandas as pd
import numpy as np

##  **Dataset 1:**

- The data contains recored values of **concentration of various pollutants** in various **cities of india** from a time span of **2015 to 2020.** The observations are recorded **daily.**

- The data also contains the Air Quality Index (AQI) value for each day along the Air Quality class (**Good, Satisfactory, Moderate, Poor, Very Poor, and Severe.**)

In [ ]:
# Dataset 1
path = 'https://gitlab.com/academics3/insaid-data/-/raw/main/Career%20Wednesday%20Repo/17th%20August%202022%3A%20Data%20Manipulation%20Assessment/Air_Quality_index.csv'

# First Look
df = pd.read_csv(path, index_col=['Date'])
df.head(7)

City       PM2.5        PM10      NO    NO2     NOx  \
Date                                                                   
2015-01-01  Ahmedabad   74.248718  178.056732    0.92  18.22   17.15   
2015-01-02  Ahmedabad   52.135330  118.105331    0.97  15.69   16.46   
2015-01-03  Ahmedabad   47.204445   88.524117   17.40  19.30   29.70   
2015-01-04  Ahmedabad   60.280689  108.053131    1.70  18.48   17.97   
2015-01-05  Ahmedabad   69.363564  129.961456   22.10  21.42   37.76   
2015-01-06  Ahmedabad  103.742111  200.129669   45.41  38.48   81.50   
2015-01-07  Ahmedabad   79.691780  234.770416  112.16  40.62  130.77   

                  NH3      CO    SO2      O3  Benzene  Toluene  Xylene  \
Date                                                                     
2015-01-01   4.478580    0.92  27.64  133.36     0.00     0.02    0.00   
2015-01-02  23.405872    0.97  24.55   34.06     3.68     5.50    3.77   
2015-01-03  11.040450   17.40  29.07   30.70     6.80    16.40    2.25   
2015-01-04  11.120313    1.70  18.59   36.08     4.43    10.14    1.00   
2015-01-05  23.415892   22.10  39.33   39.31     7.01    18.89    2.78   
2015-01-06  32.293022   45.41  45.76   46.51     5.42    10.83    1.93   
2015-01-07  36.635113  112.16  32.28   33.47     0.00     0.00    0.00   

                   AQI Air_Quality  
Date                                
2015-01-01  218.015533    Moderate  
2015-01-02  119.359024    Moderate  
2015-01-03  300.527374    Moderate  
2015-01-04  125.372055    Moderate  
2015-01-05  340.377930    Moderate  
2015-01-06  862.018616    Moderate  
2015-01-07  949.032532    Moderate

<a name = Section1></a>
## **Question 1:**

- **Industrial waste** is the combination of SO2, O3, Benzene, Toluene, and Xylene.

- **Vehicle waste** is the combination of  PM2.5, PM10, NO2, NH3, and CO.

- And **Total waste** is the **sum** of Industrial waste and vehicle waste.

- Using this information format the dataframe and find the **average total waste each city has produced.**

- Finally identify the **top 7 Most Polluted cities** based on Total waste from the obtained results.

In [ ]:
# Explore

# Creating the combined dataframe of Indutrial and Vehicle Waste
df['Industrial_Waste'] = df[['SO2', 'O3', 'Benzene', 'Toluene', 'Xylene']].sum(axis=1) #Adding all Industry related waste
df['Vehicle_Waste'] = df[['PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO']].sum(axis=1) #Adding all vehicle elated waste
df['Total_Waste'] = df[['Industrial_Waste', 'Vehicle_Waste']].sum(axis=1) #Total Pollutants in the air

# Lets take a look at combined dataframe
df.head(7)

City       PM2.5        PM10      NO    NO2     NOx  \
Date                                                                   
2015-01-01  Ahmedabad   74.248718  178.056732    0.92  18.22   17.15   
2015-01-02  Ahmedabad   52.135330  118.105331    0.97  15.69   16.46   
2015-01-03  Ahmedabad   47.204445   88.524117   17.40  19.30   29.70   
2015-01-04  Ahmedabad   60.280689  108.053131    1.70  18.48   17.97   
2015-01-05  Ahmedabad   69.363564  129.961456   22.10  21.42   37.76   
2015-01-06  Ahmedabad  103.742111  200.129669   45.41  38.48   81.50   
2015-01-07  Ahmedabad   79.691780  234.770416  112.16  40.62  130.77   

                  NH3      CO    SO2      O3  Benzene  Toluene  Xylene  \
Date                                                                     
2015-01-01   4.478580    0.92  27.64  133.36     0.00     0.02    0.00   
2015-01-02  23.405872    0.97  24.55   34.06     3.68     5.50    3.77   
2015-01-03  11.040450   17.40  29.07   30.70     6.80    16.40    2.25   
2015-01-04  11.120313    1.70  18.59   36.08     4.43    10.14    1.00   
2015-01-05  23.415892   22.10  39.33   39.31     7.01    18.89    2.78   
2015-01-06  32.293022   45.41  45.76   46.51     5.42    10.83    1.93   
2015-01-07  36.635113  112.16  32.28   33.47     0.00     0.00    0.00   

                   AQI Air_Quality  Industrial_Waste  Vehicle_Waste  \
Date                                                                  
2015-01-01  218.015533    Moderate            161.02     293.994031   
2015-01-02  119.359024    Moderate             71.56     227.736534   
2015-01-03  300.527374    Moderate             85.22     230.569011   
2015-01-04  125.372055    Moderate             70.24     219.304133   
2015-01-05  340.377930    Moderate            107.32     326.120911   
2015-01-06  862.018616    Moderate            110.45     546.964803   
2015-01-07  949.032532    Moderate             65.75     746.807309   

            Total_Waste  
Date                     
2015-01-01   455.014031  
2015-01-02   299.296534  
2015-01-03   315.789011  
2015-01-04   289.544133  
2015-01-05   433.440911  
2015-01-06   657.414803  
2015-01-07   812.557309

In [ ]:
# Lets calculate the average for each city
df_tmp = df.groupby('City').mean()
df_tmp.head()

PM2.5        PM10         NO        NO2        NOx        NH3  \
City                                                                           
Ahmedabad  66.772547  149.810513  19.554648  53.248336  43.206103  34.400908   
Aizawl     16.914135   23.200516   9.408053   0.388496  12.610354  22.308938   
Amaravati  38.754015   79.485669   4.780751  23.386241  16.380035  14.242389   
Amritsar   54.567642  115.817815  18.255660  19.307336  32.817385  15.261320   
Bengaluru  36.979296   83.316488   9.421284  28.009339  19.708455  22.674968   

                  CO        SO2         O3   Benzene    Toluene     Xylene  \
City                                                                         
Ahmedabad  15.770529  41.663039  40.816177  5.120732  24.612656   4.596121   
Aizawl      0.283628   7.378053   3.597515  9.935907  10.182015  25.012585   
Amaravati   0.643886  14.035460  38.749576  0.856810   2.676483   1.468049   
Amritsar    0.602472   8.293180  23.722732  4.268738   3.465163   6.888007   
Bengaluru   1.833318   5.532368  32.674679  3.386343   4.853583   3.899121   

                  AQI  Industrial_Waste  Vehicle_Waste  Total_Waste  
City                                                                 
Ahmedabad  360.022809        116.808726     382.763584   499.572310  
Aizawl      34.956127         56.106075      85.114120   141.220195  
Amaravati   97.427565         57.786378     177.672986   235.459363  
Amritsar   121.037974         46.637820     256.629629   303.267450  
Bengaluru   98.569488         50.346094     201.943147   252.289241

In [ ]:
# Sorting based on Total Waste
df_tmp = df_tmp.sort_values('Total_Waste', ascending=False)
df_tmp.head(7)

PM2.5        PM10         NO        NO2        NOx        NH3  \
City                                                                            
Delhi      117.112130  228.770934  38.986439  50.793259  58.567023  41.979981   
Patna      113.163608  193.673427  29.680393  37.855442  44.327395  34.543976   
Ahmedabad   66.772547  149.810513  19.554648  53.248336  43.206103  34.400908   
Gurugram   112.253185  193.805891  17.075539  24.197961  26.669894  32.190589   
Lucknow    109.636024  184.638997  15.187783  33.293691  23.800054  29.568238   
Mumbai      64.379502  151.222703  29.804524  36.090669  48.072516  27.154191   
Kolkata     64.105144  116.058727  26.568064  39.881712  63.336733  18.370381   

                  CO        SO2         O3    Benzene    Toluene    Xylene  \
City                                                                         
Delhi       1.976053  15.518474  50.394709   3.544480  17.185042  2.277337   
Patna       1.492483  20.892011  38.992118   1.847996  10.730522  5.440736   
Ahmedabad  15.770529  41.663039  40.816177   5.120732  24.612656  4.596121   
Gurugram    1.354996   9.778616  34.852053   2.770192   6.698970  3.717955   
Lucknow     2.116506  11.343699  37.045325   3.194369   5.426103  3.092843   
Mumbai      0.575263  14.216175  38.635753   1.458040   3.025343  1.843322   
Kolkata     0.799251   8.379216  30.872283  10.857506  15.377599  2.997589   

                  AQI  Industrial_Waste  Vehicle_Waste  Total_Waste  
City                                                                 
Delhi      258.639178         88.920042     538.185820   627.105861  
Patna      223.557851         77.903384     454.736723   532.640107  
Ahmedabad  360.022809        116.808726     382.763584   499.572310  
Gurugram   217.599389         57.817787     407.548055   465.365842  
Lucknow    222.438730         60.102339     398.241293   458.343633  
Mumbai     151.052824         59.178633     357.299369   416.478002  
Kolkata    139.069935         68.484193     329.120012   397.604205

In [ ]:
# Extracting Top 7 Cities
df_tmp = df_tmp.reset_index()[['City', 'Total_Waste']]
df_tmp.loc[:6]

City  Total_Waste
0      Delhi   627.105861
1      Patna   532.640107
2  Ahmedabad   499.572310
3   Gurugram   465.365842
4    Lucknow   458.343633
5     Mumbai   416.478002
6    Kolkata   397.604205

In [ ]:
# Answer
Top_7_polluted_cities = ['Delhi', 'Patna', 'Ahmedabad', 'Gurugram', 'Luckhnow', 'Mumbai', 'Kolkata']

# Checking your Answer (Don't Change this Code)
CheckingAnswer('Top7', Top_7_polluted_cities) # Delhi, Patna, Ah, Guru, Luck, Eu, Mumbai

The Given Solution is Correct

Given Soluion: ['Delhi', 'Patna', 'Ahmedabad', 'Gurugram', 'Luckhnow', 'Mumbai', 'Kolkata']
Correct Solution: ['Delhi', 'Patna', 'Ahmedabad', 'Gurugram', 'Luckhnow', 'Mumbai', 'Kolkata']


<a name = Section2></a>
## **Question 2:**

- Using the same dataset from **Question1**

- Focusing on the **Top 7 Polluted cities**, Find the number of days each of the Top 7 Polluted cities have been given **"Poor" Air Quality Class**.

- Find the **total number of days**, the data has been recorded for in dataset.

- Calculate the percentage of days the Air Quality Class was **"Poor" for each the Top 7 Most Polluted Cities.**

- Round the **percentage value to 2 decimal place** and create a dictionary with **City names as keys and percentage value as values.**

**Example:**

Taking the **Example of Delhi:**

Number of **Days Delhi had "Poor" Quality Air = 542**  
**Total number of days** in dataframe = **29531**    
**Percentage** = (542/29831) * 100 = **1.8353594**  
Round to **2 decimal**: **1.84**  
Add this to the Dictionary: {"Delhi" : 1.84}

Now, you have to do this for every city in the Top 7 Most Polluted Tips

In [ ]:
# Explore
df.head()

City      PM2.5        PM10     NO    NO2    NOx        NH3  \
Date                                                                           
2015-01-01  Ahmedabad  74.248718  178.056732   0.92  18.22  17.15   4.478580   
2015-01-02  Ahmedabad  52.135330  118.105331   0.97  15.69  16.46  23.405872   
2015-01-03  Ahmedabad  47.204445   88.524117  17.40  19.30  29.70  11.040450   
2015-01-04  Ahmedabad  60.280689  108.053131   1.70  18.48  17.97  11.120313   
2015-01-05  Ahmedabad  69.363564  129.961456  22.10  21.42  37.76  23.415892   

               CO    SO2      O3  Benzene  Toluene  Xylene         AQI  \
Date                                                                     
2015-01-01   0.92  27.64  133.36     0.00     0.02    0.00  218.015533   
2015-01-02   0.97  24.55   34.06     3.68     5.50    3.77  119.359024   
2015-01-03  17.40  29.07   30.70     6.80    16.40    2.25  300.527374   
2015-01-04   1.70  18.59   36.08     4.43    10.14    1.00  125.372055   
2015-01-05  22.10  39.33   39.31     7.01    18.89    2.78  340.377930   

           Air_Quality  Industrial_Waste  Vehicle_Waste  Total_Waste  
Date                                                                  
2015-01-01    Moderate            161.02     293.994031   455.014031  
2015-01-02    Moderate             71.56     227.736534   299.296534  
2015-01-03    Moderate             85.22     230.569011   315.789011  
2015-01-04    Moderate             70.24     219.304133   289.544133  
2015-01-05    Moderate            107.32     326.120911   433.440911

In [ ]:
df_tmp = df.reset_index()
df_tmp

Date           City      PM2.5        PM10     NO    NO2    NOx  \
0      2015-01-01      Ahmedabad  74.248718  178.056732   0.92  18.22  17.15   
1      2015-01-02      Ahmedabad  52.135330  118.105331   0.97  15.69  16.46   
2      2015-01-03      Ahmedabad  47.204445   88.524117  17.40  19.30  29.70   
3      2015-01-04      Ahmedabad  60.280689  108.053131   1.70  18.48  17.97   
4      2015-01-05      Ahmedabad  69.363564  129.961456  22.10  21.42  37.76   
...           ...            ...        ...         ...    ...    ...    ...   
29526  2020-06-27  Visakhapatnam  15.020000   50.940000   7.68  25.06  19.54   
29527  2020-06-28  Visakhapatnam  24.380000   74.090000   3.42  26.06  16.53   
29528  2020-06-29  Visakhapatnam  22.910000   65.730000   3.45  29.53  18.33   
29529  2020-06-30  Visakhapatnam  16.640000   49.970000   4.05  29.26  18.80   
29530  2020-07-01  Visakhapatnam  15.000000   66.000000   0.40  26.85  14.05   

             NH3     CO    SO2      O3    Benzene    Toluene    Xylene  \
0       4.478580   0.92  27.64  133.36   0.000000   0.020000  0.000000   
1      23.405872   0.97  24.55   34.06   3.680000   5.500000  3.770000   
2      11.040450  17.40  29.07   30.70   6.800000  16.400000  2.250000   
3      11.120313   1.70  18.59   36.08   4.430000  10.140000  1.000000   
4      23.415892  22.10  39.33   39.31   7.010000  18.890000  2.780000   
...          ...    ...    ...     ...        ...        ...       ...   
29526  12.470000   0.47   8.55   23.30   2.240000  12.070000  0.730000   
29527  11.990000   0.52  12.72   30.14   0.740000   2.210000  0.380000   
29528  10.710000   0.48   8.42   30.96   0.010000   0.010000  0.000000   
29529  10.030000   0.52   9.84   28.30   0.000000   0.000000  0.000000   
29530   5.200000   0.59   2.10   17.05  10.378896  17.546354  9.224036   

              AQI   Air_Quality  Industrial_Waste  Vehicle_Waste  Total_Waste  
0      218.015533      Moderate        161.020000     293.994031   455.014031  
1      119.359024      Moderate         71.560000     227.736534   299.296534  
2      300.527374      Moderate         85.220000     230.569011   315.789011  
3      125.372055      Moderate         70.240000     219.304133   289.544133  
4      340.377930      Moderate        107.320000     326.120911   433.440911  
...           ...           ...               ...            ...          ...  
29526   41.000000          Good         46.890000     131.180000   178.070000  
29527   70.000000  Satisfactory         46.190000     156.990000   203.180000  
29528   68.000000  Satisfactory         39.400000     151.140000   190.540000  
29529   54.000000  Satisfactory         38.140000     129.270000   167.410000  
29530   50.000000          Good         56.299286     128.090000   184.389286  

[29531 rows x 19 columns]

In [ ]:
df_tmp.groupby('Date').count()

City  PM2.5  PM10  NO  NO2  NOx  NH3  CO  SO2  O3  Benzene  \
Date                                                                     
2015-01-01     6      6     6   6    6    6    6   6    6   6        6   
2015-01-02     6      6     6   6    6    6    6   6    6   6        6   
2015-01-03     6      6     6   6    6    6    6   6    6   6        6   
2015-01-04     7      7     7   7    7    7    7   7    7   7        7   
2015-01-05     7      7     7   7    7    7    7   7    7   7        7   
...          ...    ...   ...  ..  ...  ...  ...  ..  ...  ..      ...   
2020-06-27    26     26    26  26   26   26   26  26   26  26       26   
2020-06-28    26     26    26  26   26   26   26  26   26  26       26   
2020-06-29    26     26    26  26   26   26   26  26   26  26       26   
2020-06-30    26     26    26  26   26   26   26  26   26  26       26   
2020-07-01    26     26    26  26   26   26   26  26   26  26       26   

            Toluene  Xylene  AQI  Air_Quality  Industrial_Waste  \
Date                                                              
2015-01-01        6       6    6            6                 6   
2015-01-02        6       6    6            6                 6   
2015-01-03        6       6    6            6                 6   
2015-01-04        7       7    7            7                 7   
2015-01-05        7       7    7            7                 7   
...             ...     ...  ...          ...               ...   
2020-06-27       26      26   26           26                26   
2020-06-28       26      26   26           26                26   
2020-06-29       26      26   26           26                26   
2020-06-30       26      26   26           26                26   
2020-07-01       26      26   26           26                26   

            Vehicle_Waste  Total_Waste  
Date                                    
2015-01-01              6            6  
2015-01-02              6            6  
2015-01-03              6            6  
2015-01-04              7            7  
2015-01-05              7            7  
...                   ...          ...  
2020-06-27             26           26  
2020-06-28             26           26  
2020-06-29             26           26  
2020-06-30             26           26  
2020-07-01             26           26  

[2009 rows x 18 columns]

In [ ]:
# Delhi

df[(df['City'] == 'Delhi') & (df['Air_Quality'] == 'Poor')].shape[0]

542

In [ ]:
# Empty Dictionary
Answer_Dict = {}

for city in Top_7_polluted_cities:

  # Calculating number of days each city in Top 7 Most Polluted City List
  Poor_days = df[(df['Air_Quality'] == "Poor") & (df['City'] == city)].shape[0]
  
  #Calculate Percentage
  Answer_Dict[city] = round((Poor_days/2009) * 100, 2)

In [ ]:
Answer_Dict

{'Delhi': 26.98,
 'Patna': 11.75,
 'Ahmedabad': 11.85,
 'Gurugram': 15.43,
 'Luckhnow': 0.0,
 'Mumbai': 1.74,
 'Kolkata': 5.92}

In [ ]:
df.reset_index(inplace=True)
df_polluted_days = df.groupby(['City'])['Date'].size().sort_values(ascending=False)
df_polluted_days

<a name = Section3></a>
## **Question 3:**

- Use the **same dataset**

- Two of the most influential Pollutants in the dataset are **PM2.5 and NO** for Vehicle and Industrial Waste Category Respectively.

- For **PM2.5** any concentration **equal and higher than 400** is considered Unhealthy.

- For **NO** any concentration **equal and higher than 100** is considered Unhealthy.

- Extract all the **days in the dataset** that had **Unhealthy level of PM2.5 and NO Concentration.**

- From the **Extracted data**, find the **city with highest average AQI.**

In [ ]:
# Explore
df_tmp = df[(df['PM2.5'] >= 400) & (df['NO'] >= 100)]

In [ ]:
df_tmp = df_tmp.groupby('City').mean()

df_tmp

PM2.5        PM10          NO         NO2         NOx  \
City                                                                  
Delhi    517.025714  608.444286  141.904286  121.957143  125.062857   
Lucknow  429.155000  515.668335  108.180000   68.855000  123.605000   
Patna    475.944545  387.884176  159.164545   40.075455  198.930000   

               NH3        CO        SO2         O3   Benzene    Toluene  \
City                                                                      
Delhi    91.031429  2.705714  30.647143  83.442857  9.968571  32.565714   
Lucknow  60.645000  2.180000  11.915000  30.650000  3.510000   3.625000   
Patna    61.419484  4.977273   6.707273  21.253636  0.705455   2.206364   

           Xylene         AQI  Industrial_Waste  Vehicle_Waste  Total_Waste  
City                                                                         
Delhi    9.336931  612.857143        165.961217    1608.131429  1774.092646  
Lucknow  3.008996  562.000000         52.708996    1308.288335  1360.997331  
Patna    1.953636  535.818182         32.826364    1328.395478  1361.221841

In [ ]:
df_tmp.sort_values(by='AQI', ascending=False)

PM2.5        PM10          NO         NO2         NOx  \
City                                                                  
Delhi    517.025714  608.444286  141.904286  121.957143  125.062857   
Lucknow  429.155000  515.668335  108.180000   68.855000  123.605000   
Patna    475.944545  387.884176  159.164545   40.075455  198.930000   

               NH3        CO        SO2         O3   Benzene    Toluene  \
City                                                                      
Delhi    91.031429  2.705714  30.647143  83.442857  9.968571  32.565714   
Lucknow  60.645000  2.180000  11.915000  30.650000  3.510000   3.625000   
Patna    61.419484  4.977273   6.707273  21.253636  0.705455   2.206364   

           Xylene         AQI  Industrial_Waste  Vehicle_Waste  Total_Waste  
City                                                                         
Delhi    9.336931  612.857143        165.961217    1608.131429  1774.092646  
Lucknow  3.008996  562.000000         52.708996    1308.288335  1360.997331  
Patna    1.953636  535.818182         32.826364    1328.395478  1361.221841

In [ ]:
df_tmp.index[0]

'Delhi'

In [ ]:
# Extracting days with Dangerous Level of PM2.5 and SO2 Concentration
df_tmp = df[(df['PM2.5'] >= 400) & (df['NO'] >= 100)]
df_tmp.head()

In [ ]:
# Calculating Average AQI for Each city in the extracted data
df_tmp = df_tmp.groupby('City').mean()
df_tmp

In [ ]:
# Sorting based on AQI
df_tmp = df_tmp.sort_values('AQI', ascending=False)
df_tmp[['AQI']]

In [ ]:
# Answer
Most_Unhealthy_City = 'Delhi'

# Checking your Answer (Don't Change this Code)
CheckingAnswer('Most_Unhealthy_City', Most_Unhealthy_City)

The Given Solution is Correct

Given Soluion: Delhi
Correct Solution: Delhi
